src/lang/base/Datatypes.ml

```diff
@@ -121,6 +121,20 @@ module DataTypeDictionary = struct
       adt_name_dict;
     ht
 
+  let add_adt (new_adt : adt) =
+    let open Caml in
+    match Hashtbl.find_opt adt_name_dict new_adt.tname with
+    | Some _ ->
+        fail0 @@ sprintf "Multiple declarations of type %s" new_adt.tname
+    | None ->
+        let _ = Hashtbl.add adt_name_dict new_adt.tname new_adt in
+        foldM new_adt.tconstr ~init:()
+          ~f:(fun () ctr ->
+              match Hashtbl.find_opt adt_cons_dict ctr.cname with
```

Comment:
Ah, good point. I'll add it.

Reply:
Actually, is that strictly necessary? It will cause code such as

```
type T =
| Foo

type T2 =
| T of Utin32
```
to fail, but is there any reason why we should disallow it? As far as I can tell constructors will never be confused with types by the checker or the interpreter, although it may cause some confusion for the programmer.

If we decide to disallow them, we should add a check that the contructor name isn't the same as the type that it is a constructor of, such as 

```
type T =
| T of Uint32
```

Reply:
I assumed that allowing the same names for Constructors and Types is something we don't want. I do agree that it's unlikely they'll be confused by the parser / other phases.

My take is not allow the name to be re-used, but I won't insist on it. It likely won't be a problem (as you pointed out).

Reply:
Following a discussion on slack we have decided to allow these sorts of name clashes. We already allow `Pair` as both a type name and a constructor name, so there is no reason to restrict it for custom ADTs.